@navikt/core/react/src/modal/Modal.tsx

```diff
@@ -79,111 +94,152 @@ type ModalLifecycle = {
+export const Modal = forwardRef<HTMLDialogElement, ModalProps>(
   (
     {
+      header,
       children,
       open,
-      onClose,
+      onBeforeClose,
+      onCancel,
+      width,
       className,
-      overlayClassName,
-      shouldCloseOnOverlayClick = true,
-      shouldCloseOnEsc = true,
-      closeButton = true,
-      "aria-describedby": ariaDescribedBy,
-      "aria-labelledby": ariaLabelledBy,
-      "aria-modal": ariaModal,
-      "aria-label": contentLabel,
       style,
-      parentSelector,
       ...rest
-    },
+    }: ModalProps,
     ref
   ) => {
-    const modalRef = useRef<ReactModal | null>(null);
+    const modalRef = useRef<HTMLDialogElement>(null);
     const mergedRef = useMemo(() => mergeRefs([modalRef, ref]), [ref]);
-    const buttonRef = useRef<HTMLButtonElement>(null);
-    const rootElement = useProvider()?.rootElement;
-    const appElement = useProvider()?.appElement;
+    const ariaLabelId = useId();
+    const ariaDescId = useId();
 
-    useEffect(() => {
-      appElement && Modal.setAppElement(appElement);
-    }, [appElement]);
+    if (useContext(ModalContext)) {
+      console.error("Modals should not be nested");
+    }
 
-    const onModalCloseRequest = (e) => {
-      if (shouldCloseOnOverlayClick || e.type === "keydown") {
-        onClose();
-      } else if (buttonRef.current) {
-        buttonRef.current.focus();
+    useEffect(() => {
+      if (needPolyfill && modalRef.current) {
+        dialogPolyfill.registerDialog(modalRef.current);
       }
-    };
+    }, [modalRef]);
 
-    const getParentSelector = () => {
-      if (parentSelector) {
-        return parentSelector;
+    useEffect(() => {
+      // We need to have this in a useEffect so that the content renders before the modal is displayed,
+      // and in case `open` is true initially.
+      if (modalRef.current && open !== undefined) {
+        if (open && !modalRef.current.open) {
+          modalRef.current.showModal();
+        } else if (!open && modalRef.current.open) {
+          modalRef.current.close();
+        }
       }
-      return rootElement !== undefined
-        ? () => rootElement as HTMLElement
-        : undefined;
-    };
+    }, [modalRef, open]);
+
+    useEffect(() => {
+      if (!modalRef.current) return;
+      if (open) document.body.classList.add(bodyClass); // In case `open` is true initially
+
+      const observer = new MutationObserver(() => {
+        if (modalRef.current?.open) document.body.classList.add(bodyClass);
+        else document.body.classList.remove(bodyClass);
+      });
+      observer.observe(modalRef.current, {
+        attributes: true,
+        attributeFilter: ["open"],
+      });
+      return () => {
+        observer.disconnect();
+        document.body.classList.remove(bodyClass); // In case modal is unmounted before it's closed
+      };
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [modalRef]);
```

Comment:
Siden useEffect er etter first-mount tror jeg man bare kan sjekke ref-en
```suggestion
    }, [modalRef, open]);

    useEffect(() => {
      if (!modalRef.current) return;
      if (modalRef.current?.open) document.body.classList.add(bodyClass); // In case `open` is true initially

      const observer = new MutationObserver(() => {
        if (modalRef.current?.open) document.body.classList.add(bodyClass);
        else document.body.classList.remove(bodyClass);
      });
      observer.observe(modalRef.current, {
        attributes: true,
        attributeFilter: ["open"],
      });
      return () => {
        observer.disconnect();
        document.body.classList.remove(bodyClass); // In case modal is unmounted before it's closed
      };
    }, [modalRef]);
```

Reply:
Godt poeng